torch/utils/benchmark/utils/timer.py

```diff
@@ -21,6 +22,57 @@ def timer() -> float:
     timer = timeit.default_timer
 
 
+PYTHON = "Python"
+CPP = "C++"
+
+
+class CPPTimer:
+    def __init__(
+        self,
+        stmt: str,
+        setup: str,
+        timer: Callable[[], float],
+        globals: Dict[str, Any],
+    ) -> None:
+        if timer is not timeit.default_timer:
+            # FIXME: We should be able to automatically detect GPU operations.
+            raise ValueError(
+                "When timing C++, timer must be `timeit.default_timer`. If you "
+                "did not specify `timer=...`, this means that a timer with "
+                "CUDA syncronization was chosen, and you should pass "
+                "`timer=timeit.default_timer` to the Timer's constructor to "
+                "declare that CUDA syncronization is not needed."
```

Comment:
*synchronization (misspelled in two places)

But why would C++ imply that you don't need CUDA synchronization? If I have a C++ binary that does GPU operations, I still need to sync to get accurate numbers.

Reply:
Sorry. I'm really bad about misspelling that.

It's doing the opposite; it's saying that it doesn't support CUDA yet. (There doesn't seem to be an analog to `torch.cuda. synchronize` in the C++ API.) Long term we should just be able to detect CUDA operations during the warmup and choose the timer appropriately, but that's for a future PR.

Reply:
(1) There should be a `torch::cuda::synchronize`; the functions in `torch::cuda` have always been missing bits and pieces here and there and could use some love. (2) It's really easy to do synchronize in C++. Here's the implementation of the Python binding of `_cuda_synchronize`:

```
PyObject * THCPModule_cudaSynchronize(PyObject *_unused, PyObject *noargs)
{
  HANDLE_TH_ERRORS
  c10::cuda::device_synchronize();
  Py_RETURN_NONE;
  END_HANDLE_TH_ERRORS
}

```

Reply:
My concern was replicating `_lazy_init()`, which is not so simple.